File: leetcode/200/LC239MaxSlidingWindow.java
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;

public class LC239MaxSlidingWindow {
    public static void main(String[] args) {
        int[] nums = new int[]{1, 3, -1, -3, 5, 3, 6, 7};
        System.out.println(Arrays.toString(new LC239MaxSlidingWindow().maxSlidingWindow(nums,3)));
    }
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> dq = new ArrayDeque<>();
        int[] ret = new int[nums.length - k + 1];
        for (int i = 0; i < nums.length; i++) {
            while (!dq.isEmpty() && nums[dq.peekLast()] <= nums[i]) {
                dq.removeLast();
            }
            while (!dq.isEmpty() && (i - dq.getFirst()) >= k) {
                dq.removeFirst();
            }

            dq.addLast(i);
            if (i+1 >= k) {
                ret[i+1-k] = nums[dq.peekFirst()];
            }
        }

        return ret;
    }
}

File: leetcode/200/lc239maxSlidingWindow.go
func maxSlidingWindow(nums []int, k int) []int {
	n := len(nums)
	dq := []int{}
	ret := make([]int, n-k+1)

	for i := 0; i < n; i++ {
		cur := nums[i]
		for len(dq) > 0 && nums[dq[len(dq)-1]] <= cur {
			dq = dq[:len(dq)-1]
		}

		for len(dq) > 0 && i-dq[0]+1 > k {
			dq = dq[1:]
		}

		dq = append(dq, i)

		if i+1 >= k {
			ret[i+1-k] = nums[dq[0]]
		}
	}

	return ret
}